docs_src/design/adr/014-Secret-Provider-For-All.md

```diff
@@ -0,0 +1,409 @@
+
+type InsecureSecretsInfo struct {
+	Path    string
+	Secrets map[string]string
+}
+```
+
+```toml
+[Writable.InsecureSecrets]
+    [Writable.InsecureSecrets.DB]
+    	path = "redisdb"
+    	[Writable.InsecureSecrets.DB.Secrets]
+            username = ""
+            password = ""
+    [Writable.InsecureSecrets.mqtt]
+    	path = "mqtt"
+    	[Writable.InsecureSecrets.mqtt.Secrets]
+            username = ""
+            password = ""
+            cacert = ""
+            clientcert = ""
+            clientkey = ""
+```
+
+## Decision
+
+The new `SecretProvider` abstraction defined by this ADR is a combination of the two implementations described above in the [Existing Implementations](#existing-implementations) section.
+
+### Only Exclusive Secret Stores
+
+To simplify the `SecretProvider` abstraction, we need to reduce to using only exclusive `SecretStores`. This allows all the APIs to deal with a single `SecretClient`, rather than the split up way we currently have in Application Services. This requires that the current Application Service shared secrets (database credentials) must be copied into each Application Service's exclusive `SecretStore` when it is created.
+
+The challenge is how do we seed static secrets for unknown services when they become known.  As described above in the [Known and Unknown Services](#known-and-unknown-services) section above,  services currently identify themselves for exclusive `SecretStore` creation via the `ADD_SECRETSTORE_TOKENS` environment variable on security-secretstore-setup. This environment variable simply takes a comma separated list of service names.
+
+```yaml
+ADD_SECRETSTORE_TOKENS: "<service-name1>,<service-name2>"
+```
+
+If we expanded this to add an optional list of static secret identifiers for each service, i.e.  `appservice/redisdb`, the exclusive store could also be seeded with a copy of static shared secrets. In this case the Redis database credentials for the Application Services' shared database. The environment variable name will change to `ADD_SECRETSTORE` now that it is more than just tokens.
+
+```yaml
+ADD_SECRETSTORE: "app-service-xyz[appservice/redisdb]"
+```
+
+> *Note: The secret identifier here is the short path to the secret in the existing **appservice**  `SecretStore`. In the above example this expands to the full path of `/secret/edgex/appservice/redisdb`*
+
+The above example results in the Redis credentials being copied into app-service-xyz's `SecretStore` at `/secret/edgex/app-service-xyz/redis`.
+
+Similar approach could be taken for Message Bus credentials where a common `SecretStore` is created with the Message Bus credentials saved. The services request the credentials are copied into their exclusive `SecretStore` using `common/messagebus` as the secret identifier.
+
+Full specification for the environment variable's value is a comma separated list of service entries defined as:
+
+```
+<service-name>[optional list of static secret IDs sperated by ;]>
```

Comment:
- syntax-wise, there is an extra `>` at the end of this example
- should have another `service-name` in this example to fit the aforementioned sentence for `comma separated list`

```suggestion
<service-name1>[optional list of static secret IDs sperated by ;],<service-name2>[optional list of static secret IDs sperated by ;],...
```

Reply:
I Like it!